### DISCUSSION TITLE
JsonPath filter expressions?

### DISCUSSION THREAD

**Discussion by svenXY** [mistweaverco/kulala.nvim#269] [Q&A] [answered]
Hi, just a quick one: does the jsonpath implementation support jsonpath filter expressions (e.g. $.results[?(@.desc == 'terra')].id) or do I have to use jq then?

```json
{
  "links": [
    {
      "href": "https://cloud.mongodb.com/",
      "rel": "self"
    }
  ],
  "results": [
    {
      "desc": "terra",
      "id": "641dcxxxxxxx",
      "links": [
        {
          "href": "https://cloud.mongodb.com/api/atlas/v2/",
          "rel": "self"
        }
      ],
      "publicKey": "foofoofoo"
    },
    {
      "desc": "svh_owner",
      "id": "63f5c4dzzzzzzzzzzzz",
      "links": [
        {
          "href": "https://clo",
          "rel": "self"
        }
      ],
      "privateKey": "********-****-****-******",
      "publicKey": "xwvzzzzzzzzz"
    }
  ],
  "totalCount": 3
}
```

and then 

`$.results[?(@.desc == 'terra')].id `
would return 641dcxxxxxxx

**Comment by Grueslayer** [comment, marked as answer]
Hi,

no. The implemented parser is - as written - a simple one where you only have access to the fields and arrays by index.

`JQ` can be used to parse but is another syntax not compatible with JsonPath.

If you would like to use complex JsonPath expression you can use [jsonpath-cli](https://www.jsware.io/jsonpath-cli/). Sadly it can not return the raw value, it will always return the value in JSON format (like `"content"`) and adds linefeeds (because it could be a list).
But we can easily write a lua wrapper for it:

```lua
  {
    'mistweaverco/kulala.nvim',
    config = function()
      require('kulala').setup({
        contenttypes = {
          ["application/json"] = {
            pathresolver = function(body, path)
              local cmd = { "jpp", "-e", path}
              local out = vim.system(cmd, { stdin = body, text = true }):wait().stdout
              out = vim.fn.json_decode(out)
              return out[1]
            end,
          },
        },
      })
    end,
  },
```

(Please use `jpp.cmd` when on MS Win)

**Reply by svenXY** [reply]
thanks for the clarification and suggestion. I might try - but not on my company laptop.

**Comment by Grueslayer** [comment]
@gorillamoe Maybe we can put this in the FAQ or the Configuration docs.

**Reply by gorillamoe** [reply]
We might also consider adding a second alternative path-resolver like the one you posted and add a setup function to it, so that it checks for for node and mpm and installs the package locally if not already installed. 

Then we could add that alternative parser with the requirements of nodes to the readme. 🤔

Any thoughts on this? Overkill or okayish?

/edit

In fact I would check for the jpp command first, if it fails then npm and node and then install, otherwise just continue

**Comment by gorillamoe** [comment]
Somehow related: https://github.com/mistweaverco/kulala.nvim/discussions/270#discussioncomment-10882171
